abyfy_errors.go

```diff
@@ -0,0 +1,104 @@
+package httpapi
+
+import (
+	"errors"
+	"strings"
+
+	"github.com/ipfs/go-cid"
+	ipld "github.com/ipfs/go-ipld-format"
+)
+
+type prePostWrappedNotFoundError struct {
+	pre  string
+	post string
+
+	wrapped ipld.ErrNotFound
+}
+
+func (e prePostWrappedNotFoundError) String() string {
+	return e.Error()
+}
+
+func (e prePostWrappedNotFoundError) Error() string {
+	return e.pre + e.wrapped.Error() + e.post
+}
+
+func (e prePostWrappedNotFoundError) Unwrap() error {
+	return e.wrapped
+}
+
+func abyfyIpldNotFoundFallbackToMSG(msg string) error {
+	err, handled := abyfyIpldNotFound(msg)
+	if handled {
+		return err
+	}
+
+	return errors.New(msg)
+}
+
+func abyfyIpldNotFoundFallbackToError(msg error) error {
+	err, handled := abyfyIpldNotFound(msg.Error())
+	if handled {
+		return err
+	}
+
+	return msg
+}
+
+// This file handle parsing and returning the correct ABI based errors from error messages
+//lint:ignore ST1008 this function is not using the error as a mean to return failure but it massages it to return the correct type
+func abyfyIpldNotFound(msg string) (error, bool) {
```

Comment:
This seems pretty complicated and brittle. Does every impl of this HTTP client have to do this? Should we consider changing the API to return structured errors, or do this server side, so clients don't have to resort to these contortions?

Reply:
Some summarizing from the issue as discussed in https://discord.com/channels/806902334369824788/956547968385822730/957037148853395496.

> IIUC from the interface tests previously we did have some weak contract on the CoreAPI that functions like Block().Get() and Block().Rm() also needed to return things like blockstore.ErrNotFound  (or really that the error message would be something like blockstore: block not found https://github.com/ipfs/interface-go-ipfs-core/commit/01ee9419a28353cab04979f0791133df9869f30a. Now we've broken the weak contract by changing the text and are figuring out the best way to restore it.

> Does every impl of this HTTP client have to do this?

Not necessarily, it depends what they're trying to do. In this case to for an HTTP API to be compatible with `interface-go-ipfs-core`'s CoreAPI interfaces it needs to do some sort of error checking to figure out what type of error it received. Maybe we could drop that requirement, but idk that we have to.

> Should we consider changing the API to return structured errors

You're right that it might be worth changing the HTTP API to have some contract on the types of errors it returns. I think for me the question boils down to the value tradeoffs here, i.e. the number of server and client implementations of this API that need to care about errors vs how much effort it is to return structured errors. Given the low number of servers and the low number of clients that care strongly about the error types it seems like we can probably not do this right now.

> or do this server side, so clients don't have to resort to these contortions?

We can make some of this easier server side, although currently there are two implementers of this API (go-ipfs and js-ipfs) with most other IPFS implementations choosing to not implement this HTTP API at all. Some contortions will still be necessary because go-ipfs and js-ipfs don't have the same error messages at the moment (especially since we just changed them in go-ipfs with resolving https://github.com/ipfs/go-ipfs/issues/7074).

Reply:
OK thanks. For reference, here are some active community client impls (other than Go and JS): 

* https://github.com/ipfs-shipyard/py-ipfs-http-client
* https://github.com/ferristseng/rust-ipfs-api
* https://github.com/vasild/cpp-ipfs-http-client
* https://github.com/Peergos/Peergos/blob/00e2a8d121473002fa9bdc1ca3f90ecf7f9d33e8/src/peergos/shared/storage/ContentAddressedStorage.java#L257

I don't know how these are used and whether they care about this or not.

Reply:
> I don't know how these are used and whether they care about this or not.

My thought is that grep.app would've shown more code caring about the previous error text if code built around those clients cared. Since there doesn't seem to be much there it appears to be safe to change the error text and that others aren't relying on knowing the error types.

Reply:
> For reference, here are some active community client impls

I don't think it matters to them, as this is required if the http clients want to be compatible with the CoreAPI objects (extended weak contract, if you only use errors to show your users error messages you don't care about that for example).

Object which doesn't exists in other languages and even if it does (*maybe JS ?*) it doesn't know about `ipld.ErrNotFound`.